java/sage/io/EncryptedSageFile.java

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2015 The SageTV Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sage.io;
+
+import sage.IOUtils;
+
+import java.io.IOException;
+
+// If you want to add more encryption implementations, be sure to take a look at SageDataFile
+// initialization as it depends on this class name to determine when a source is encrypted.
+public class EncryptedSageFile implements SageFileSource
+{
+  private final byte[] cryptoKeys;
+  private final SageFileSource sageFileSource;
+  // This is used and increased in size when required to encrypt bytes so we are not encrypting the
+  // bytes in the original byte array.
+  private byte writeBuffer[];
+
+  /**
+   * Create a new encryption filter over a <code>SageFileSource</code>
+   * <p/>
+   * This will seamlessly encrypt and decrypt reads and writes done through this class.
+   * <p/>
+   * The default key will be used.
+   *
+   * @param sageFileSource The <code>SageFileSource</code> to wrap inside this filter.
+   * @throws IOException If there is an I/O related error or the <code>SageFileSource</code> is
+   *                     another instance of <code>EncryptedSageFile</code>.
+   */
+  public EncryptedSageFile(SageFileSource sageFileSource) throws IOException
+  {
+    this(sageFileSource, IOUtils.getCryptoKeys());
+  }
+
+  /**
+   * Create a new encryption filter over a <code>SageFileSource</code>
+   * <p/>
+   * This will seamlessly encrypt and decrypt reads and writes done through this class.
+   *
+   * @param sageFileSource The <code>SageFileSource</code> to wrap inside this filter.
+   * @param crypt Specify alternative 128 byte key.
+   * @throws IOException If there is an I/O related error or the <code>SageFileSource</code> is
+   *                     another instance of <code>EncryptedSageFile</code>.
+   */
+  public EncryptedSageFile(SageFileSource sageFileSource, byte crypt[]) throws IOException
+  {
+    // In case we try to make a file that's already using this "filter" use it twice in a row.
+    if (sageFileSource instanceof EncryptedSageFile)
```

Comment:
I suppose the way this is built that's a real possibility that one could be buried in the stack. This is here to help out SageDataFile so it doesn't need to be aware of multiple encryption layers. Either way, I made the search recursive per your suggestion.